File: Chapter02/Exercise2_03/WetSchemaParser.scala
package Exercise2_03

import Utilities02.WetRecord
import org.apache.hadoop.io.Text
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import Utilities01.HelperScala.createSession
import Utilities02.HelperScala.{sampleWetLoc, extractRawRecords, parseRawWet}

/**
 * Code for parsing .wet files of the WARC corpus
 *
 * @author Phil, https://github.com/g1thubhub
 */
object WetSchemaParser {

  def main(args: Array[String]) = {

    implicit val session: SparkSession = createSession(3, "WET Parser")
    session.sparkContext.setLogLevel("ERROR") // avoids printing of info messages

    val rawRecords: RDD[Text] = extractRawRecords(sampleWetLoc)
    val wetRecords: RDD[WetRecord] = rawRecords
      .flatMap(parseRawWet)

    import session.implicits._
    wetRecords.toDF().printSchema()
    println(s"Total # of records: ${wetRecords.count()}")

  }
}

File: Chapter02/Exercise2_03/wet_schema_parser.py
from Chapter01.utilities01_py.helper_python import create_session
from Chapter02.utilities02_py.helper_python import sample_wet_loc, extract_raw_records, parse_raw_wet

if __name__ == "__main__":
    session = create_session(3, 'WET Parser')
    session.sparkContext.setLogLevel('ERROR')  # avoids printing of info messages

    raw_records = extract_raw_records(sample_wet_loc, session)
    wet_records = raw_records.flatMap(lambda record: parse_raw_wet(record))

    wet_records.toDF().printSchema()
    print('Total # of records: ' + str(wet_records.count()))

